scylla-cql/src/types/deserialize/value.rs

```diff
@@ -984,6 +984,70 @@ where
     }
 }
 
+// tuples
+
+// Implements tuple deserialization.
+// The generated impl expects that the serialized data contains exactly the given amount of values.
+macro_rules! impl_tuple {
+    ($($Ti:ident),*; $($idx:literal),*; $($idf:ident),*) => {
+        impl<'frame, $($Ti),*> DeserializeValue<'frame> for ($($Ti,)*)
+        where
+            $($Ti: DeserializeValue<'frame>),*
+        {
+            fn type_check(typ: &ColumnType) -> Result<(), TypeCheckError> {
+                const TUPLE_LEN: usize = (&[$($idx),*] as &[i32]).len();
+                let [$($idf),*] = ensure_tuple_type::<($($Ti,)*), TUPLE_LEN>(typ)?;
+                $(
+                    <$Ti>::type_check($idf)?;
+                )*
+                Ok(())
+            }
+
+            fn deserialize(typ: &'frame ColumnType, v: Option<FrameSlice<'frame>>) -> Result<Self, DeserializationError> {
+                const TUPLE_LEN: usize = (&[$($idx),*] as &[i32]).len();
+                // Safety: we are allowed to assume that type_check() was already called.
+                let [$($idf),*] = ensure_tuple_type::<($($Ti,)*), TUPLE_LEN>(typ)
+                    .expect("Type check should have prevented this!");
+
+                // Ignore the warning for the zero-sized tuple
+                #[allow(unused)]
+                let mut v = ensure_not_null_frame_slice::<Self>(typ, v)?;
+                let ret = (
+                    $(
+                        <$Ti>::deserialize($idf, v
+                            .read_cql_bytes()
+                            .map_err(|err| mk_deser_err::<Self>(
+                                typ,
+                                TupleDeserializationErrorKind::ElementDeserializationFailed {
+                                    err: DeserializationError::new(err),
+                                    index: TUPLE_LEN - 1 - $idx
+                                }
+                            ))?
+                        )?,
+                    )*
+                );
+                Ok(ret)
+            }
+        }
+    }
+}
+
+macro_rules! impl_tuple_multiple {
+    (;< $($T_PREF:ident,)* >;;< $($idx_pref:literal,)* >;;< $($idf_pref:ident,)* >) => {
+        impl_tuple!(;;);
+    };
+    ($TN:ident $(,$Ti:ident)*; < $($T_PREF:ident,)* > ;  $idx_n:literal $(,$idx:literal)*; < $($idx_pref:literal,)* >; $idf_n:ident $(,$idf:ident)* ; <$($idf_pref:ident,)*>) => {
+        impl_tuple!($($T_PREF,)* $TN ; $($idx_pref, )* $idx_n; $($idf_pref, )* $idf_n);
+        impl_tuple_multiple!($($Ti),* ; < $($T_PREF,)* $TN, > ; $($idx),*; < $($idx_pref, )* $idx_n,  >; $($idf),*; < $($idf_pref, )* $idf_n, >);
+    }
+}
```

Comment:
This macro needs some explanation about how it works, it probably won't be easy to understand for a new contributor.

Reply:
For writing a comment explaining how it works, first I have to remember and understand it... :P